### PR TITLE
Add check for not valid constraints

### PR DESCRIPTION
**Pull request by BLoHny**
original: #374 
Relates to https://github.com/mfvanek/pg-index-health/issues/362

### CODE REVIEW

File: pg-index-health/src/main/java/io/github/mfvanek/pg/checks/host/NotValidConstraintsOnHost.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    @Nonnull
+    @Override
+    public List<Constraint> check(@Nonnull final PgContext pgContext) {
+        return executeQuery(pgContext, rs -> {
+            final String tableName = rs.getString(TABLE_NAME);
+            final String constraintName = rs.getString("constraint_name");
+            final String constraintType = rs.getString("constraint_type");
+            final ConstraintType ct = constraintType.equals("c") ? ConstraintType.CHECK : ConstraintType.FOREIGN_KEY;
```
Comment (mfvanek) [inline pg-index-health/src/main/java/io/github/mfvanek/pg/checks/host/NotValidConstraintsOnHost.java:35]: Use `ConstraintType.valueOf()`

File: pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/Constraint.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import io.github.mfvanek.pg.model.DbObject;
+import io.github.mfvanek.pg.model.table.TableNameAware;
+import io.github.mfvanek.pg.model.validation.Validators;
+
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.Immutable;
+import java.util.Objects;
+
+@Immutable
+public class Constraint implements DbObject, TableNameAware {
```
Comment (mfvanek) [inline pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/Constraint.java:22]: Add javadoc for all new classes

File: pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/Constraint.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    private final String tableName;
+    private final String constraintName;
+    private final ConstraintType constraintType;
+
+    private Constraint(
```
Comment (mfvanek) [inline pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/Constraint.java:28]: Mark as `protected`

File: pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/Constraint.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    @Nonnull
+    public ConstraintType getConstraintType() {
+        return constraintType;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) return true;
```
Comment (mfvanek) [inline pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/Constraint.java]: Please format code as others https://github.com/mfvanek/pg-index-health/blob/328192f97307781c00fc6543a362e01d9c2f071c/pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/ForeignKey.java#L90

Comment (mfvanek) [reply pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/Constraint.java]: Mark method `final`

File: pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/Constraint.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+        Constraint that = (Constraint) o;
+
+        if (!tableName.equals(that.tableName)) return false;
+        if (!constraintName.equals(that.constraintName)) return false;
+        return constraintType == that.constraintType;
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (mfvanek) [inline pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/Constraint.java:76]: Mark method `final`

File: pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/Constraint.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        return Objects.hash(tableName, constraintName, constraintType);
+    }
+
+    @Override
+    public String toString() {
+        return "Constraint{" +
```
Comment (mfvanek) [inline pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/Constraint.java:82]: Use `.class.getSimpleName()` here https://github.com/mfvanek/pg-index-health/blob/328192f97307781c00fc6543a362e01d9c2f071c/pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/ForeignKey.java#L119C26-L119C48

File: pg-index-health/src/main/java/io/github/mfvanek/pg/checks/host/NotValidConstraintsOnHost.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import io.github.mfvanek.pg.common.maintenance.Diagnostic;
+import io.github.mfvanek.pg.connection.PgConnection;
+import io.github.mfvanek.pg.model.PgContext;
+import io.github.mfvanek.pg.model.constraint.Constraint;
+import io.github.mfvanek.pg.model.constraint.ConstraintType;
+
+import java.util.List;
+import javax.annotation.Nonnull;
+
+public class NotValidConstraintsOnHost extends AbstractCheckOnHost<Constraint> {
```
Comment (mfvanek) [inline pg-index-health/src/main/java/io/github/mfvanek/pg/checks/host/NotValidConstraintsOnHost.java:22]: Please rename to `NotValidConstraintsCheckOnHost`.  
See https://github.com/mfvanek/pg-index-health/pull/374#discussion_r1581845619

File: pg-index-health/src/main/java/io/github/mfvanek/pg/checks/cluster/NotValidConstraintsOnCluster.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+package io.github.mfvanek.pg.checks.cluster;
+
+import io.github.mfvanek.pg.checks.host.NotValidConstraintsOnHost;
+import io.github.mfvanek.pg.connection.HighAvailabilityPgConnection;
+import io.github.mfvanek.pg.model.constraint.Constraint;
+
+import javax.annotation.Nonnull;
+
+public class NotValidConstraintsOnCluster extends AbstractCheckOnCluster<Constraint> {
```
Comment (mfvanek) [inline pg-index-health/src/main/java/io/github/mfvanek/pg/checks/cluster/NotValidConstraintsOnCluster.java:19]: Please rename to `NotValidConstraintsCheckOnCluster`.   
See https://github.com/mfvanek/pg-index-health/pull/374#discussion_r1581845740

File: pg-index-health/src/test/java/io/github/mfvanek/pg/checks/host/NotValidConstraintsOnHostTest.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        executeTestOnDatabase(schemaName, dbp -> dbp.withNotValidConstraints().withUniqueConstraintOnSerialColumn(), ctx ->
+                assertThat(check)
+                        .executing(ctx)
+                        .hasSize(2)
+                        .containsExactly(
+                                Constraint.of(ctx.enrichWithSchema("accounts"), "c_accounts_chk_client_id_not_validated_yet",
+                                        ConstraintType.CHECK),
+                                Constraint.of(ctx.enrichWithSchema("accounts"), "c_accounts_fk_client_id_not_validated_yet",
+                                        ConstraintType.FOREIGN_KEY)
+                        )
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/checks/host/NotValidConstraintsOnHostTest.java:49]: Add `validate constraint` call and after that recheck again (should be empty).  
See https://github.com/mfvanek/pg-index-health/pull/374#issuecomment-2081961766

For example
```sql
ALTER TABLE distributors VALIDATE CONSTRAINT distfk;
```
https://www.postgresql.org/docs/16/sql-altertable.html

File: pg-index-health/src/test/java/io/github/mfvanek/pg/checks/cluster/NotValidConstraintsOnClusterTest.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+                    .containsExactly(
+                            Constraint.of(ctx.enrichWithSchema("accounts"), "c_accounts_chk_client_id_not_validated_yet",
+                                    ConstraintType.CHECK),
+                            Constraint.of(ctx.enrichWithSchema("accounts"), "c_accounts_fk_client_id_not_validated_yet",
+                                    ConstraintType.FOREIGN_KEY)
+                    );
+
+            assertThat(check.check(ctx, FilterTablesByNamePredicate.of(ctx.enrichWithSchema("accounts"))))
+                    .isEmpty();
+        });
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/checks/cluster/NotValidConstraintsOnClusterTest.java:52]: Add validate constraint call and after that recheck again (should be empty).

File: pg-index-health/src/test/java/io/github/mfvanek/pg/checks/cluster/NotValidConstraintsOnClusterTest.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+import io.github.mfvanek.pg.model.constraint.ConstraintType;
+import io.github.mfvanek.pg.support.DatabaseAwareTestBase;
+import io.github.mfvanek.pg.support.DatabasePopulator;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class NotValidConstraintsOnClusterTest extends DatabaseAwareTestBase {
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/checks/cluster/NotValidConstraintsOnClusterTest.java:27]: Please rename test class the same as corresponding check-class

File: pg-index-health/src/test/java/io/github/mfvanek/pg/checks/host/NotValidConstraintsOnHostTest.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+import io.github.mfvanek.pg.model.constraint.Constraint;
+import io.github.mfvanek.pg.model.constraint.ConstraintType;
+import io.github.mfvanek.pg.support.DatabaseAwareTestBase;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import static io.github.mfvanek.pg.support.AbstractCheckOnHostAssert.assertThat;
+
+public class NotValidConstraintsOnHostTest extends DatabaseAwareTestBase {
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/checks/host/NotValidConstraintsOnHostTest.java:25]: Please rename test class the same as corresponding check-class

**Review by mfvanek — CHANGES_REQUESTED** [review]
Good job! Just need to polish

File: pg-index-health/src/test/java/io/github/mfvanek/pg/checks/host/NotValidConstraintsOnHostTest.java (hunk 9 above)
Comment (BLoHny) [reply pg-index-health/src/test/java/io/github/mfvanek/pg/checks/host/NotValidConstraintsOnHostTest.java:49]:
```java

//example..
        statement.execute(String.format("alter table %1$s.accounts validate constraint c_accounts_fk_client_id_not_validated_yet;",
                schemaName));
        statement.execute(String.format("alter table %1$s.accounts validate constraint c_accounts_chk_client_id_not_validated_yet;",
                schemaName));
```

I think it's okay to create a new statement that calls the valid constraint and test it. What are your thought?

Comment (mfvanek) [reply pg-index-health/src/test/java/io/github/mfvanek/pg/checks/host/NotValidConstraintsOnHostTest.java:49]: > I think it's okay to create a new statement that calls the valid constraint and test it. What are your thoughts?

I'd use option with `ExecuteUtils.executeOnDatabase(getDataSource(), statement -> statement.execute("vacuum analyze"));`
https://github.com/mfvanek/pg-index-health/blob/master/pg-index-health/src/test/java/io/github/mfvanek/pg/support/ExecuteUtils.java#L29

Comment (BLoHny) [reply pg-index-health/src/test/java/io/github/mfvanek/pg/checks/host/NotValidConstraintsOnHostTest.java:49]:
```java
    @ParameterizedTest
    @ValueSource(strings = {PgContext.DEFAULT_SCHEMA_NAME, "custom"})
    void onDatabaseWithThem(final String schemaName) {
        executeTestOnDatabase(schemaName, dbp -> dbp.withNotValidConstraints().withUniqueConstraintOnSerialColumn(), ctx -> {
            assertThat(check)
                    .executing(ctx)
                    .hasSize(2)
                    .containsExactly(
                            Constraint.of(ctx.enrichWithSchema("accounts"), "c_accounts_chk_client_id_not_validated_yet", ConstraintType.CHECK),
                            Constraint.of(ctx.enrichWithSchema("accounts"), "c_accounts_fk_client_id_not_validated_yet", ConstraintType.FOREIGN_KEY)
                    );

            ExecuteUtils.executeOnDatabase(DatabaseAwareTestBase.getDataSource(), statement -> {
                statement.execute(String.format("alter table %1$s.accounts validate constraint c_accounts_fk_client_id_not_validated_yet;", schemaName));
                statement.execute(String.format("alter table %1$s.accounts validate constraint c_accounts_chk_client_id_not_validated_yet;", schemaName));
            });

            assertThat(check)
                    .executing(ctx)
                    .isEmpty();
        });
    }
```

proceed like this?

Comment (mfvanek) [reply pg-index-health/src/test/java/io/github/mfvanek/pg/checks/host/NotValidConstraintsOnHostTest.java:49]: > proceed like this?

Yeah!

**Review by mfvanek — CHANGES_REQUESTED** [review]
There are failed tests `352 tests completed, 13 failed`
and issues with code coverage
```
* What went wrong:
Execution failed for task ':pg-index-health-model:jacocoTestCoverageVerification'.
> Rule violated for bundle pg-index-health-model: classes missed count is 2.0, but expected maximum is 0.0
  Rule violated for bundle pg-index-health-model: methods missed count is 13.0, but expected maximum is 0.0
  Rule violated for bundle pg-index-health-model: lines missed count is 30.0, but expected maximum is 0.0
  Rule violated for bundle pg-index-health-model: instructions covered ratio is 0.9, but expected minimum is 1.0
  Rule violated for bundle pg-index-health-model: branches covered ratio is 0.8, but expected minimum is 1.0
```

**Comment by BLoHny** [comment]
Even though I filled the test coverage of the newly created class to 100% and corrected other test errors, the coverage is not filled due to classes other than the coverage stack. What should I do?

**Comment by mfvanek** [comment]
> Even though I filled the test coverage of the newly created class to 100% and corrected other test errors, the coverage is not filled due to classes other than the coverage stack. What should I do?
> 
> ` branches covered ratio is 0.9, but expected minimum is 1.0`

JaCoCo provides an html report. You can analyze it and figure out parts not covered properly with tests.  
I need to see your latest changes in order to give more feedback.

P.S. Also you can split you PR into two parts and push changes to model separately
I mean pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/Constraint.java
pg-index-health-model/src/main/java/io/github/mfvanek/pg/model/constraint/ConstraintType.java

**Comment by BLoHny** [comment]
https://github.com/mfvanek/pg-index-health/blob/master/pg-index-health-model/src/testFixtures/java/io/github/mfvanek/pg/support/TestUtils.java

image


This is where the coverage issue for the private constructor. I don't think test coverage for this class is needed. What are your thoughts?

**Comment by mfvanek** [comment]
> This is where the coverage issue for the private constructor. I don't think test coverage for this class is needed. What are your thoughts?

It's definitely not an issue. We don't collect code coverage from test classes. Push your changes and I'll take a look

**Comment by mfvanek** [comment]
Hi @BLoHny,  
I cherry-picked your commits and made some improvements.  
Your code is already in the master branch.

**Pull request closed without merging**